src/REP15.sol

```diff
@@ -0,0 +1,476 @@
+
+      if (readyForDetachmentAt == 0) {
+        revert REP15NotRequestedForDetachment(ctxHash, tokenId);
+      }
+
+      if (readyForDetachmentAt > block.timestamp) {
+        revert REP15UnreadyForDetachment(ctxHash, tokenId, uint64(block.timestamp), readyForDetachmentAt);
+      }
+    }
+
+    delete _tokenContext[tokenId][ctxHash];
+    _removeAttachedContext(tokenId, ctxHash);
+
+    if (emitEvent) emit ContextDetached(ctxHash, tokenId);
+
+    address controller = _contexts[ctxHash].controller;
+    if (controller.code.length > 0) {
+      try IREP15ContextCallback(controller).onExecDetachContext(ctxHash, tokenId, operator, data) { } catch { }
+    }
+  }
+
+  /**
+   * @dev Overrides the internal `_update` function to revoke ownership delegation and detach all attached contexts
+   * in case of transfers or burns.
+   *
+   * If `auth` is non-zero and this function will check if `auth` is the ownership manager, an authorized operator of
+   * ownership manager, or the approved address for this NFT (if the token is not being delegated).
+   */
+  function _update(address to, uint256 tokenId, address auth) internal virtual override returns (address) {
+    if (auth != address(0)) {
+      _checkAuthorizedOwnershipManager(tokenId, auth);
+    }
+
+    if (auth != address(0) || to == address(0)) {
+      // Revoke current ownership delegation if any.
+      // No need to check if the delegation is active or emit the OwnershipDelegationStopped event.
+      delete _delegations[tokenId];
+
+      // Detach all attached contexts. No need to emit the ContextDetached event.
+      bytes32[] storage attachedContexts = _attachedContexts[tokenId];
+      for (int256 i = int256(attachedContexts.length) - 1; i >= 0; --i) {
+        bytes32 ctxHash = attachedContexts[uint256(i)];
+        _detachContext(ctxHash, tokenId, auth, "", _tokenContext[tokenId][ctxHash].locked, false);
+      }
+    }
+
+    return super._update(to, tokenId, address(0));
+  }
+
+  /**
+   * @dev Checks if the `delegatee` is the owner or an approved operator of the `tokenId`.
+   */
+  function _checkAuthorizedDelegatee(address delegatee, address operator, uint256 tokenId) internal view virtual {
+    if (!(delegatee == operator || isApprovedForAll(delegatee, operator))) {
+      revert REP15InsufficientApproval(operator, tokenId);
+    }
+  }
+
+  /**
+   * @dev Checks if the `operator` is the ownership manager, an authorized operator of ownership manager,
+   * or the approved address for this NFT (if the token is not being delegated).
+   */
+  function _checkAuthorizedOwnershipManager(uint256 tokenId, address operator) internal view virtual {
+    REP15Utils.Delegation storage $delegation = _delegations[tokenId];
+
+    if (!$delegation.isActive()) {
+      ERC721._checkAuthorized(_ownerOf(tokenId), operator, tokenId);
+      return;
+    }
+
+    _checkAuthorizedDelegatee($delegation.delegatee, operator, tokenId);
```

Comment:
I think you can rename the function `_checkAuthorizedDelegatee` to something like `_checkOwnerOperator` and write this function as follow:
```suggestion
    address manager = getOwnershipManager(tokenId);
    _checkOwnerOperator(manager, operator, tokenId);
```

Reply:
That is not the desired behavior.
Recall:
> Checks if the `operator` is the ownership manager, an authorized operator of the ownership manager,  or the approved address for this NFT (if the token is not being delegated).

So your code (`_checkOwnerOperator`) still needs to check whether `operator` is the approved address (via sing-token approval) if the token is not being delegated. In this case, your code is similar to mine: you must check `!$delegation.isActive()` and call `ERC721._checkAuthorized`.

Reply:
yeah I forgot the case where token is not delegated which the token approval must be allowed